### ISSUE TITLE
Option for showing open window previews on click action.

### ISSUE THREAD

**Issue by Amplificator** [micheleg/dash-to-dock#465]
Now that we have the "Show open windows previews" when we right click on an icon, it would be great if we could have it go straight to the "All Windows" section to immediately show the window previous.
And if we could set that as the left mouse click action that would be awesome :)

**Comment by franglais125** [comment]
I think this is a duplicate of #418.

Anyway, I made a first implementation here: https://github.com/franglais125/dash-to-dock/tree/click_window_preview

**Comment by Amplificator** [comment]
Indeed, for the most part. Difference seems to be that I requested that it could be chosen to happen on left-click.

**Comment by franglais125** [comment]
Yes, it's getting a bit hard, as right-click, left-click and hover are possible options. Maybe a duplicate of this one https://github.com/micheleg/dash-to-dock/issues/418 in that case?

If you want left click, check out this branch: https://github.com/franglais125/dash-to-dock/tree/click_window_preview

I'm still trying to make it nicer, but it works already.

**Comment by Amplificator** [comment]
I get an error on that one.
Removed the old one, restarted Gnome, cloned the branch, restarted gnome again.

I get this:
GLib.FileError: Failed to open file '/home/martin/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/schemas/gschemas.compiled': open() failed: No such file or directory

```
Stack trace:
  getSettings@/home/martin/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/convenience.js:64
  Settings<._init@/home/martin/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/prefs.js:58
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178
  _Base.prototype._construct@resource:///org/gnome/gjs/modules/lang.js:110
  Class.prototype._construct/newClass@resource:///org/gnome/gjs/modules/lang.js:213
  buildPrefsWidget@/home/martin/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/prefs.js:675
  Application<._selectExtension@resource:///org/gnome/shell/extensionPrefs/main.js:90
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178
  Application<._onCommandLine@resource:///org/gnome/shell/extensionPrefs/main.js:239
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178
  main@resource:///org/gnome/shell/extensionPrefs/main.js:377
  @<main>:1
```

**Comment by franglais125** [comment]
Did you install with `make && make install`?

What you are having is essentially a missing option in the schemas, which means it wasn't recompiled.

**Comment by Amplificator** [comment]
Yeah, I forgot that, already fixed that though, but the window preview doesn't happen here.. the behavior is the same as when choosing to cycle between windows.

**Comment by Amplificator** [comment]
Basically it should open the right-click menu and already have opened the "All Windows" tab.

**Comment by franglais125** [comment]
Did you select the option in the settings?

![screenshot from 2017-04-07 18-48-10](https://cloud.githubusercontent.com/assets/19195975/24822392/ce1dd0b8-1bc2-11e7-9b7e-51fc16ced0b0.png)

**Comment by Amplificator** [comment]
Yes :)

**Comment by franglais125** [comment]
I'm a bit puzzled... It should work for icons with more than 1 window.

![screenshot from 2017-04-07 18-58-43](https://cloud.githubusercontent.com/assets/19195975/24822594/42c4b84a-1bc4-11e7-8251-fa8f3f3ce537.png)

**Comment by Amplificator** [comment]
It manages to show 1½ preview image and then it freezes up Gnome completely.

**Comment by franglais125** [comment]
Well, sorry to hear that. :(

What version of Gnome-Shell do you use? And do you get any error messages that can help debug?

**Comment by Amplificator** [comment]
Gnome 3.22.3 running in Arch.

No errors, it just locks up :/

**Comment by franglais125** [comment]
@Amplificator here is another branch, with a different implementation https://github.com/franglais125/dash-to-dock/tree/click_window_preview_fromPopupMenu

This one actually opens the popupMenu (same as right click), but forces the window preview section to be open.

**Comment by Amplificator** [comment]
@franglais125 that branch works flawlessly, thank you :)

Would you mind making a pull request to have it merged to the official plugin?

**Comment by franglais125** [comment]
@Amplificator I actually don't like the latest branch I linked, as it feels more hacky: I'm opening the right-click menu and setting a variable to keep it forcibly open.

The first solution was cleaner, while introducing more code. I now logged in into a X session and I can reproduce the freeze. I had no problems under Wayland!

In the end, I'll try to come up with a better solution, from the point of view of the code, and that doesn't crash your shell ;)

**Comment by franglais125** [comment]
@Amplificator would you mind trying again the first branch? I pushed a fix which hopefully addresses your issue.
https://github.com/franglais125/dash-to-dock/tree/click_window_preview

**Comment by Amplificator** [comment]
@franglais125 This fix seems to work :)

I actually like this version better because it only shows images and none of the buttons like the 2nd branch you made did.

Can it be made to not show above the main Gnome bar? When not transparent it hides whatever is on the bar. If it started just below it I think it would look better.

![screenshot](http://i.imgur.com/oT3Y3Zk.png)

Don't mind the semi-transparentness of it, that is not an issue but something I did to take the screenshot.

**Comment by franglais125** [comment]
From #504:
> This was merged into master [0772814]..[3c9f7f3].
